pyanaconda/modules/security/security.py

```diff
@@ -176,15 +180,64 @@ def set_realm(self, realm):
         self.realm_changed.emit()
         log.debug("Realm is set to %s.", realm)
 
+    def handle_realm_discover_results(self, results):
+        """ Handle results from the RealmDiscover task.
+
+        :param results: a two element results tuple returned by the task,
+                        the first element is a boolean tracking if the discovery was successful,
+                        the second element is a list of packages required to be installed on the
+                        target system to successfully join the realm
+        """
+        log.debug("Handling results from realm discover task.")
+        realm_discovered, required_packages = results
+
+        # record if the discovery was successful
+        realm_data = self.realm
+        realm_data.realm_discovered = realm_discovered
+        self.set_realm(realm_data)
+
+        # record package requirements
+        self._realm_required_packages = required_packages
+
+    def collect_requirements(self):
+        """Return installation requirements for this module.
+
+        :return: a list of requirements
+        """
+        requirements = []
+
+        # Add realm requirements.
+        for name in self._realm_required_packages:
+            requirements.append(Requirement.for_package(name, reason="realm join"))
+
+        return requirements
+
+    def setup_with_tasks(self):
+        """Return the setup tasks for this module."""
+        realm_task = RealmDiscoverTask(sysroot=conf.target.system_root,
+                                      realm_name=self.realm.name,
+                                      discover_options=self.realm.discover_options)
+
+        realm_task.succeeded_signal.connect(lambda: self.handle_realm_discover_results(realm_task.get_result()))
+
+        tasks = [realm_task]
+
+        paths = [
+            self.publish_task(SECURITY.namespace, task) for task in tasks
+        ]
+
+        return paths
 
     def install_with_tasks(self):
         """Return the installation tasks of this module.
 
         :returns: list of object paths of installation tasks
         """
-        tasks = [
-            ConfigureSELinuxTask(sysroot=conf.target.system_root, selinux_mode=self.selinux)
-        ]
+        tasks = [ConfigureSELinuxTask(sysroot=conf.target.system_root, selinux_mode=self.selinux),
+                 RealmJoinTask(sysroot=conf.target.system_root,
+                               realm_discovered=self.realm.realm_discovered,
+                               realm_name=self.realm.name,
+                               join_options=self.realm.join_options)]
```

Comment:
It looks like `RealmJoinTask` could get just the system root and  `self.realm`.

Reply:
Yep, this should work as long as we update the realm data when realm_changed signal is triggered. :)